Improved SDA target (#51)

jdhoffa opened the pull request:
- [x] This PR needs some SERIOUS refactoring, but it works. I have added a ref-file which @maurolepore you can use if you want as a reference to refactor right now. I didn't think of any more interesting tests yet, so that's sorta all we got to stick with. I will continue working on this on Friday after the holiday tomorrow.
- [x] This PR replaces the currently existing files "R/sda_portfolio_target.R", "market.R" and "portfolio.R", which should then be removed in a later PR. 
- [x] This PR has temporary commits that should be moved to ".data". All these commits relate to the demo data file `co2_intensity_scenario`

Have a look and let me know what you think, I know it's a bit of a monster so we will likely need to work on it for a while.

I am also going to post a reprex that might be able to give you some context into what the desired output is. 

On the plus side, WE'RE ALMOST READY TO GET .anlysis on CRAN! :D

jdhoffa commented:
``` r
library(r2dii.data)
library(r2dii.match)
library(r2dii.analysis)
library(ggplot2)

valid_matches <- match_name(loanbook_demo, ald_demo) %>%
  # WARNING: Remember to validate matches (see `?prioritize`)
  prioritize()

sda_targets <- valid_matches %>%
  add_sda_target(
    ald = ald_demo,
    co2_intensity_scenario = r2dii.analysis::co2_intensity_scenario
  )


ggplot(sda_targets) + 
  geom_line(aes(year, emission_factor_value, colour = emission_factor_name))
```

![](https://i.imgur.com/8WSRk6s.png)

<sup>Created on 2020-05-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

jdhoffa commented:
I see R CMD CHECK is failing on CI, which confuses me. All checks pass locally for me. 
That said I'm calling it a day now, so I will leave that to you to debug ;)

maurolepore commented:
The Details of the check logs give some hints. You may need to `spelling::update_wordslist()`, and to also check for problems in the examples section. You may run `devtools::run_examples(fresh =  TRUE)` to iterate faster than R CMD check.

maurolepore commented:
I see 

```r
 Error: Column `production_weighted_emission_factor` must be length 1 (a summary value), not 32
```

I can reproduce it with dplyr 0.8.5

``` r
devtools::load_all()
#> Loading r2dii.analysis
packageVersion("dplyr")
#> [1] '0.8.5'

installed <- requireNamespace("r2dii.data", quietly = TRUE) &&
  requireNamespace("r2dii.match", quietly = TRUE)
if (!installed)
  stop("Please install r2dii.match and r2dii.data")

library(r2dii.data)
library(r2dii.match)

valid_matches <- match_name(loanbook_demo, ald_demo) %>%
  # WARNING: Remember to validate matches (see `?prioritize`)
  prioritize()

valid_matches %>%
  add_sda_target(
    ald = ald_demo,
    co2_intensity_scenario = r2dii.analysis::co2_intensity_scenario
  )
#> Error: Column `production_weighted_emission_factor` must be length 1 (a summary value), not 32
```

<sup>Created on 2020-05-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

maurolepore commented:
I  confirm that using the dplyr ~1 fixes CI. I'll now try fix the code so it also works with dplyr < 1

maurolepore commented:
And here is the gist of the problem and solution

``` r
library(dplyr, warn.conflicts = FALSE)
packageVersion("dplyr")
#> [1] '0.8.5'
library(tidyr)

# Fails
out <- mtcars %>% 
  as_tibble() %>% 
  summarise(
    # Output has length = 1
    cyl = sum(cyl), 
    # Outputs more than a sigle value
    disp = range(disp)
  )
#> Error: Column `disp` must be length 1 (a summary value), not 2

# Passes
out <- mtcars %>% 
  as_tibble() %>% 
  summarise(
    # Output has length = 1
    cyl = sum(cyl), 
    # Output has length = 1
    disp = list(
      # Output has length > 1
      range(disp)
    )
  )

out %>% unnest(cols = disp)
#> # A tibble: 2 x 2
#>     cyl  disp
#>   <dbl> <dbl>
#> 1   198  71.1
#> 2   198 472
```

<sup>Created on 2020-05-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

maurolepore commented on DESCRIPTION:
```diff
@@ -43,7 +43,8 @@ Imports:
     rlang (>= 0.1.2),
-    tibble
+    tibble,
+    tidyr
```
We need `tidyr::unnest()`

maurolepore commented on NAMESPACE:
```diff
@@ -18,17 +19,25 @@ export(summarize_portfolio_production)
 export(summarize_weighted_production)
 export(sym)
 export(syms)
+importFrom(dplyr,arrange)
```
Imported functions from dplyr to reduce "noise" so I can see what matters more clearly.

maurolepore commented on R/add_company_target.R:
```diff
@@ -39,7 +39,7 @@ add_company_target <- function(data) {
   initial_sector_summaries <- data %>%
     dplyr::group_by(!!!rlang::syms(by_company)) %>%
-    dplyr::summarise(
+    dplyr::summarize(
```
For consistency

maurolepore commented on R/add_sda_target.R:
```diff
@@ -0,0 +1,185 @@
+#' Add portfolio level SDA target
```
Haven't read the documentation yet. Maybe Vincent would be a good person to review it.

maurolepore commented on R/add_sda_target.R:
```diff
@@ -0,0 +1,185 @@
+  start_year <- co2_intensity_scenario %>%
+    select(.data$sector, .data$year) %>%
+    group_by(.data$sector) %>%
+    summarize(start_year = min(.data$year))
```
Every time we call `summarize()` with a function `f()` we need to think and test how `f()` handles `NA`. For example, `min()`  defaults to `na.rm = TRUE`; better to be explicit to about this -- it may cause errors that we need to test for.

maurolepore commented on R/add_sda_target.R:
```diff
@@ -0,0 +1,185 @@
+    select(-.data$start_year)
+
+
+  ald_sda_market_benchmark <- calculate_sda_market_benchmark(
```
Maybe remove the prefix `calculate_`. The pro of making it a verb is outweighed by the con of a function name that is probably too long. Long names are generally not a problem but they may obscure the intent of the code.

maurolepore commented on R/add_sda_target.R:
```diff
@@ -0,0 +1,185 @@
+  market %>%
+    group_by(.data$sector, .data$year) %>%
+    summarize(
+      sector_total_production = sum(.data$production),
```
Did you think about missing values?

maurolepore commented on R/add_sda_target.R:
```diff
@@ -0,0 +1,185 @@
+    summarize(
+      sector_total_production = sum(.data$production),
+      # Alias production_weighted_emission_factor
+      x. = list(.data$production * .data$emission_factor)
```
`list()` and `unnest()` are needless in dplyr >= 1 but best to keep them for backward compatibility.

maurolepore commented on R/add_sda_target.R:
```diff
@@ -0,0 +1,185 @@
+    arrange(.data$year) %>%
+    mutate(
+    # styler: off
+      x. =  .data$emission_factor,  # For short so the pattern is clearer
+      g =   .data$x. / first(.data$x.),
+      py = (.data$x. -  last(.data$x.)) / (first(.data$x.) - last(.data$x.)),
+      x. = NULL
```
Feel free to undo this. I just wanted to see more clearly what's going on.

maurolepore commented on R/add_sda_target.R:
```diff
@@ -0,0 +1,185 @@
+    distinct(.data$id_loan, .data[[loan_size]])
+
+  total_size_by_sector <- distinct_loans_by_sector %>%
+    summarize(total_size = sum(.data[[loan_size]]))
```
Watch out for `NA`s.

maurolepore commented on R/add_sda_target.R:
```diff
@@ -0,0 +1,185 @@
+    add_weighted_loan_emission_factor(use_credit_limit = use_credit_limit) %>%
+    group_by(.data$sector, .data$year, ...) %>%
+    summarize(
+      weighted_emission_factor = sum(.data$weighted_loan_emission_factor)
```
Watch out for `NA`s.

maurolepore commented on inst/WORDLIST:
```diff
@@ -1,11 +1,14 @@
 DegreesInvesting
 dfrac
 dii
+DII
```
Are we supposed to use DII or Dii (uppercase or lowercase i)?

jdhoffa replied:
technically uppercase, 2DII, but people jump around a lot

jdhoffa commented:
- [ ] Draft an article explaining the SDA calculation 
- [ ] Write an explicit expected output test, and use it to guide refactoring
- [ ] How should we handle NAs from input data?
- [ ] Try to break the code, and write corresponding tests

jdhoffa commented:
@maurolepore 
* I added a vignette explaining the rational behind this target calculation
* I added several typical tests to see how the code handles bad input data
* I added one pretty strict output test (c2d18bf) to hopefully give us a bit more confidence refactoring
* I did a small bit of refactoring

I still need to look into how the code handles `NA`s in more detail. 
When you start tomorrow, could you move the `r2dii.data` related commits over to that repo in a PR?

jdhoffa commented:
e41dd4e 
232f819
1131e84

jdhoffa commented:
I cherry-picked data-related commits and moved them to .data. Once 2DegreesInvesting/r2dii.data#83 is merged, we can revert the following commits: 
e41dd4e
232f819
1131e84
827ac05
20f5f1b
c1eba6b
0220c8a

jdhoffa commented on R/add_sda_target.R:
```diff
@@ -46,6 +46,7 @@ add_sda_target <- function(data,
     "loan_size_outstanding",
     "loan_size_credit_limit",
     "name_ald",
+    "sector",
```
This seems like it shouldn't be crucial to me. I'll need to explore this a little bit.

maurolepore commented:
Not sure why my regression references didn't work and yours did. I removed all references and re-built them. I think we had a similar problem before in r2dii.data -- except that it was my references that worked and yours that failed. I behaves as if gh actions had a cache of the references that wouln't update -- but I can't see how the references would be cached. ASAIK the workflow we use creates a cache only of packages.

jdhoffa commented:
That's very peculiar...

jdhoffa commented:
Ok sorted it. No issue in `join_ald_scenario`. Neither function errors if you drop the `sector` column of valid_matches, see reprex below: 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(r2dii.data)
library(r2dii.match)
library(r2dii.analysis)

valid_matches <- match_name(loanbook_demo, ald_demo) %>%
  # WARNING: Remember to validate matches (see `?prioritize`)
  prioritize()

valid_matches %>%
  dplyr::select(-sector) %>% 
  join_ald_scenario(
    ald = ald_demo,
    scenario = scenario_demo_2020,
    region_isos = region_isos_demo
  )
#> # A tibble: 9,945 x 43
#>    id_loan id_direct_loant… name_direct_loa… id_intermediate… name_intermedia…
#>    <chr>   <chr>            <chr>            <chr>            <chr>           
#>  1 L151    C168             Shaanxi Auto     <NA>             <NA>            
#>  2 L151    C168             Shaanxi Auto     <NA>             <NA>            
#>  3 L151    C168             Shaanxi Auto     <NA>             <NA>            
#>  4 L151    C168             Shaanxi Auto     <NA>             <NA>            
#>  5 L151    C168             Shaanxi Auto     <NA>             <NA>            
#>  6 L151    C168             Shaanxi Auto     <NA>             <NA>            
#>  7 L151    C168             Shaanxi Auto     <NA>             <NA>            
#>  8 L151    C168             Shaanxi Auto     <NA>             <NA>            
#>  9 L151    C168             Shaanxi Auto     <NA>             <NA>            
#> 10 L151    C168             Shaanxi Auto     <NA>             <NA>            
#> # … with 9,935 more rows, and 38 more variables: id_ultimate_parent <chr>,
#> #   name_ultimate_parent <chr>, loan_size_outstanding <dbl>,
#> #   loan_size_outstanding_currency <chr>, loan_size_credit_limit <dbl>,
#> #   loan_size_credit_limit_currency <chr>, sector_classification_system <chr>,
#> #   sector_classification_input_type <chr>,
#> #   sector_classification_direct_loantaker <dbl>, fi_type <chr>,
#> #   flag_project_finance_loan <chr>, name_project <lgl>,
#> #   lei_direct_loantaker <lgl>, isin_direct_loantaker <lgl>, id_2dii <chr>,
#> #   level <chr>, sector_ald <chr>, name <chr>, name_ald <chr>, score <dbl>,
#> #   source <chr>, technology <chr>, production_unit <chr>, year <int>,
#> #   production <dbl>, emission_factor <dbl>, country_of_domicile <chr>,
#> #   plant_location <chr>, number_of_assets <int>, is_ultimate_owner <lgl>,
#> #   is_ultimate_listed_owner <lgl>, ald_timestamp <chr>,
#> #   ald_emission_factor_unit <chr>, scenario <chr>, region <chr>, tmsr <dbl>,
#> #   smsp <dbl>, scenario_source <chr>

valid_matches %>%
  dplyr::select(-sector) %>% 
  add_sda_target(
    ald = ald_demo,
    co2_intensity_scenario = co2_intensity_scenario_demo
  )
#> # A tibble: 28 x 4
#> # Groups:   sector [1]
#>    sector  year emission_factor_name               emission_factor_value
#>    <chr>  <dbl> <chr>                                              <dbl>
#>  1 cement  2020 portfolio_weighted_emission_factor                 0.664
#>  2 cement  2020 portfolio_target_emission_factor                   0.669
#>  3 cement  2020 scenario_emission_factor                           0.7  
#>  4 cement  2021 portfolio_weighted_emission_factor                 0.665
#>  5 cement  2021 portfolio_target_emission_factor                   0.612
#>  6 cement  2021 scenario_emission_factor                           0.64 
#>  7 cement  2022 portfolio_weighted_emission_factor                 0.666
#>  8 cement  2022 portfolio_target_emission_factor                   0.555
#>  9 cement  2022 scenario_emission_factor                           0.580
#> 10 cement  2023 portfolio_weighted_emission_factor                 0.667
#> # … with 18 more rows
```

<sup>Created on 2020-05-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

maurolepore commented on R/add_sda_target.R:
```diff
@@ -1,7 +1,7 @@
-#' This function calculates the portfolio-level emissions factor targets using
-#' the sectoral-decarbonization approach.
+#' This function calculates emissions-factor targets at the portfolio-level. It
+#' uses the sectoral-decarbonization approach.
```
Do we have a link with details about what is the sectoral-decarbonization approach?

jdhoffa replied:
I think the vignette/ article I wrote is the best bet

maurolepore commented on R/add_sda_target.R:
```diff
@@ -1,6 +1,6 @@
 #' Add targets of CO2 emissions at the portolio level, using the SDA approach
 #'
-#' This function calculates emissions-factor targets at the portfolio-level. It
+#' This function calculates targets of CO2 emissions at the portfolio-level. It
```
I removed the word 'factor' from 'emissions-factor' because it means nothing to me, and instead I added the word 'CO2". This may be inaccurate but is clearer to me. Undo this if most readers should immediately know what 'emission-factors' mean; else I suggest you keep it like this or include an explanation (or a link to it).

jdhoffa replied:
It doesn’t calculate emissions, the targets are for the “emissions factors” which is short for “emissions intensity factor” or “CO2 emissions per unit production”. Not sure if you prefer any of those synonyms, but it’s important to specify that we aren’t speaking about absolute emissions

maurolepore replied:
What I find more meaningful is "CO2 emissions per unit production". I understand every word in isolation and they make sense together as a phrase.

I find the word "factor" too vague. It could mean a million things unless it is defined immediately before the first time you use it.

In the article you wrote:
* I fail to match any reference to "emissions factor".
* I do match "emission intensity"  in "The company-level emission intensity target is then defined as ..."
* I also match "unit of production" in "Hence to reflect this difference, a rate of change, $g(t)$ is taken from the average sector CO2-intensity per unit of production in a specific year ..."

jdhoffa commented:
@maurolepore trying to decide how to approach this emission_factor terminology issue. I understand the conundrum from a description point of view, but I think the jargon most used/ understood by the users will be "emission factor". Also, in the ald_data, scenario_data, documentation, they are sure to see the words emission factor somewhere (for example: http://data.iea.org/payment/products/122-emissions-factors-2017-edition.aspx) 

So from this point of view, i don't want to change how we refer to is in the actual code (aka i want the column names to be `emission_factor_*`) but I'm happy to expand on the explanation in either the function description, the vignette, or both. What do you think?

maurolepore commented:
That sounds like a great plan. As long as the newbies like me have a way to learn the jargon from the docs, then the jargon is good as it communicates complex concepts succinctly.
